File: frontend/src/main/frontend/lesusscrofa/src/app/model/order-summary.ts
import { Formula } from './formula';

export class OrderSummary {

    public static fromJson(json: Object) {
        return new OrderSummary(json['description'],
                                json['formula'],
                                json['unitPrice'],
                                json['vat'],
                                json['quantity'],
                                json['reduction'],
                                json['unitPriceReductionIncluded'],
                                json['unitPriceVatIncluded'],
                                json['totalVatExcluded'],
                                json['totalVatIncluded'])
    }

    constructor(private _description: string,
                private _formula: Formula,
                private _unitPrice: number,
                private _vat: number,
                private _quantity: number,
                private _reduction: number,
                private _unitPriceReductionIncluded: number,
                private _unitPriceVatIncluded: number,
                private _totalVatExcluded: number,
                private _totalVatIncluded: number) {

    }

    public get description(): string {
        return this._description;
    }

    public get formula(): Formula {
        return this._formula;
    }

    public get unitPrice(): number {
        return this._unitPrice;
    }

    public get vat(): number {
        return this._vat;
    }

    public get quantity(): number {
        return this._quantity;
    }

    public get reduction(): number {
        return this._reduction;
    }

    public get unitPriceReductionIncluded(): number {
        return this._unitPriceReductionIncluded;
    }

    public get unitPriceVatIncluded(): number {
        return this._unitPriceVatIncluded;
    }

    public get totalVatExcluded(): number {
        return this._totalVatExcluded;
    }

    public get totalVatIncluded(): number {
        return this._totalVatIncluded;
    }
}

File: api/src/main/java/be/susscrofa/api/model/OrderSummary.java
package be.susscrofa.api.model;


import lombok.*;

import java.math.BigDecimal;
import java.math.RoundingMode;

@Data
@Builder
@AllArgsConstructor
@NoArgsConstructor
public class OrderSummary {

    private long id;

    private String name;

    private Formula formula;

    private BigDecimal unitPrice;

    private int vat;

    private int quantity;

    private int reduction;

    private String unit;

    private BigDecimal computeUnitPriceReductionIncluded() {
        return unitPrice.multiply(BigDecimal.valueOf(100 - reduction).divide(BigDecimal.valueOf(100)));
    }

    public BigDecimal getUnitPriceReductionIncluded() {
        return computeUnitPriceReductionIncluded()
                .setScale(2, RoundingMode.HALF_UP);
    }

    public BigDecimal computeUnitPriceVat() {
        return computeUnitPriceReductionIncluded()
                .multiply(BigDecimal.valueOf(vat).divide(BigDecimal.valueOf(100)));
    }

    public BigDecimal computeUnitPriceVatIncluded() {
        return computeUnitPriceReductionIncluded()
                .multiply(BigDecimal.valueOf(1).add(BigDecimal.valueOf(vat).divide(BigDecimal.valueOf(100))));
    }

    public BigDecimal getUnitPriceVatIncluded() {
        return computeUnitPriceVatIncluded()
                .setScale(2, RoundingMode.HALF_UP);
    }

    private BigDecimal computeTotalVatExcluded() {
        return computeUnitPriceReductionIncluded()
                .multiply(BigDecimal.valueOf(quantity));
    }

    public BigDecimal getTotalVatExcluded() {
        return computeTotalVatExcluded()
                .setScale(2, RoundingMode.HALF_UP);
    }

    private BigDecimal computeTotalVat() {
        return computeUnitPriceVat()
                .multiply(BigDecimal.valueOf(quantity));
    }

    public BigDecimal getTotalVat() {
        return computeTotalVat()
                .setScale(2, RoundingMode.HALF_UP);
    }

    private BigDecimal computeTotalVatIncluded() {
        return computeTotalVatExcluded()
                .multiply(BigDecimal.valueOf(1).add(BigDecimal.valueOf(vat).divide(BigDecimal.valueOf(100))));
    }

    public BigDecimal getTotalVatIncluded() {
        return computeTotalVatIncluded()
                .setScale(2, RoundingMode.HALF_UP);
    }

    public String getDescription() {
        StringBuilder description = new StringBuilder();

        description.append(formula == Formula.OTHER ? name : formula.getName());
        description.append(" à ");
        description.append(getUnitPriceReductionIncluded() + " €");

        return description.toString();
    }

    public OrderSummary reduction(int reduction) {
        this.reduction = reduction;

        return this;
    }
}
